honeybee/aperture.py

```diff
@@ -0,0 +1,506 @@
+
+    @property
+    def area(self):
+        """The area of the aperture."""
+        return self._geometry.area
+
+    @property
+    def perimeter(self):
+        """The perimeter of the aperture."""
+        return self._geometry.perimeter
+
+    @property
+    def type(self):
+        """Object for Type of Aperture (ie. Window)."""
+        return self._type
+
+    @type.setter
+    def type(self, value):
+        assert value in self.TYPES, '{} is not a valid face type.'.format(value)
+        self._type = value
+
+    @property
+    def boundary_condition(self):
+        """Boundary condition of this aperture."""
+        return self._boundary_condition
+
+    @boundary_condition.setter
+    def boundary_condition(self, value):
+        assert isinstance(value, (Outdoors, Surface)), \
+            'Aperture only supports Outdoor or Surface boundary condition. ' \
+            'Got {}'.format(type(value))
+        self._boundary_condition = value
+
+    @property
+    def parent(self):
+        """Parent Face if assigned. None if not assigned."""
+        return self._parent
+
+    @property
+    def has_parent(self):
+        """Boolean noting whether this Aperture has a parent Face."""
+        return self._parent is not None
+
+    def set_adjacency(self, other_aperture):
+        """Set this aperture to be adjacent to another.
+
+        Note that this method does not verify whether the other_aperture geometry is
+        co-planar or compatible with this one so it is recommended that a test
+        be performed before using this method in order to verify these criteria.
+        The Face3D.is_centered_adjacent() or the Face3D.is_geometrically_equivalent()
+        methods are both suitable for this purpose.
+
+        Args:
+            other_aperture: Another Aperture object to be set adjacent to this one.
+        """
+        assert isinstance(other_aperture, Aperture), \
+            'Expected Aperture. Got {}.'.format(type(other_aperture))
+        self.boundary_condition = Surface(other_aperture)
+        other_aperture.boundary_condition = Surface(self)
+
+    def overhang(self, depth, angle=0, indoor=False):
+        """Get a single overhang for this Aperture.
+
+        Args:
+            depth: A number for the overhang depth.
+            angle: A number for the for an angle to rotate the overhang in degrees.
+                Default is 0 for no rotation.
+            indoor: Boolean for whether the overhang should be generated facing the
+                opposite direction of the aperture normal (typically meaning
+                indoor geometry). Default: False.
+
+        Returns:
+            overhang: A Shade object for the Aperture overhang. Will be None if the
+                Aperture geometry does not have a flat top.
+        """
+        overhang = self.generate_louvers_by_number(
+            1, depth, angle=angle, indoor=indoor)
+        return overhang[0] if len(overhang) != 0 else None
```

Comment:
Do we want to support adding overhangs, fins, etc to an aperture as shadings objects? Or are you keeping that for `honeybee-energy`? I think that will be a useful property.

Reply:
I don't think implementing the IDF equivalents of these objects is nearly as useful as having them in the core for a couple of reasons:
1. It they only exist in honeybee-energy, they become more difficult to coordinate with daylight, radiation, and thermal comfort studies that use Radiance. Here in the core,  both energy and radiance reference the same geometry.
2. Simply having the overhang represented by a depth (as it is in E+) and not having the detailed geometry information for the resulting overhang makes it more difficult to check that the overhang is modeled as intended.  For example, if the window doesn't have a perfectly flat top, you don't get a complete overhang and you want to check/visualize this geometry rather then worrying about how E+ interpreted your inputs.  There's no question about what the geometry is when you have detailed shading geometry.

Reply:
If it wasn't clear I'm in favor of having them in the core library but this comment wasn't about that. What I meant is to add a property like `shading` to `aperture` to collect shading objects so you can do something like `aperture.shading` and it returns all the fins, overhangs and the rest of the shading objects. Right now you have to collect these objects in a separate list somewhere.

Reply:
Sorry that I misunderstood.  I was considering before this but it adds another level of complexity that I am not sure is helpful because people can already assign shades to zones.  If we allow people to assign shades to apertures, it also seems weird to not let them assign shades to faces, which means that we now have to keep track of shades assigned to 4 objects (Aperture, Face, Room, Model) instead of just 2 (Room and Model).

Also, I think the Room level is really where we want people to assign them since aperture-level shading really doesn't have a meaning for E+.  Plus I am sure that certain overhangs are big enough that you really want to account for the fact that they shade neighboring windows in the room or have an effect beyond just that window.
If you want something like this that auto-assigns shades to an object, we can add some methods on the Room to generate shades across all of the windows of a certain orientation or something like that.

Reply:
This logic works well for EnergyPlus but it is not how shadings are designed most of the times and also is not how it will be used in a daylight model. I think we have to at least keep track of them in 3 objects. We can remove the assignment for Face because it will be the same as assigning them to zones in both energy and daylight models.

Here is how I would categorize them.

1. One can add shadings to Aperture (which is the most common case for blinds, fins, etc) and what will be used for dynamic daylight modeling. This will be added as a Zone shading in energy models.
2. If not a specific aperture but a specific Zone then it can be assigned to the Zone. This will be used as exterior shades in daylight and is the ideal case for energy models.
3. Model level shading will be only for context shading. Will be treated the same in both daylight and energy models.

Not to make it more complicated but there will be a need for dynamic shading objects which will be defined differently in energy versus daylight models. That can probably be handled by `honeybee-energy` and `honeybee-daylight`.
 
If you think it will be too much to ask for this PR (which I would agree) we can open an issue to discuss this further in the future.

Reply:
Ok, I'm just glad to hear you agree that it should not be a part of this PR since I don't think that I'm really be best one to implement it.  You have a much clearer idea of why you need this functionality for Radiance than I do and, as I said previously, we really don't need/want it for energy or thermal comfort. So please feel free to open an issue if you want to tackle this later and here's a list of things that should be addressed when you add support for aperture-specific shading:

1. Make it clear to the user that aperture-level shading only applies to a single aperture in Radiance (and not the apertures that are right next to it) but it applies to all apertures of the Room for energy/thermal comfort.
2. Give some warnings in the docstrings that Aperture-level overhangs and fins can be dangerous when you have two apertures that are really close to one another.  For example, a fin that looks like it is in between two apertures that are very close to one another is actually only shading one of the two apertures.
3. Have clear explanations for why the API is not consistent in how it deals with shades (ie. they can be applied to Models, Rooms, and Apertures but not Faces).  And, even though you can still generate shades from Face geometry, you cannot assign it to the Face.
4. Handle the fact that it is not clear whether Aperture-level shading is on the inside or the outside of a Room.  For example, you might generate shades for an aperture without a parent and the shades look like they are on a certain side of the aperture.  Then, the aperture gets added to a Room and the aperture face gets flipped to ensure that it is facing outwards from the Room.  This reverses the inside/outside relationship of the shades assigned to the aperture.

Reply:
I opened an issue with link to these comments and leave the conversation for the future.

https://github.com/ladybug-tools/honeybee-core/issues/11